File: packages/evals/tasks/bench/extract/extract_staff_members.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";
import { normalizeString } from "../../../framework/textScoring.js";

export default defineBenchTask(
  { name: "extract_staff_members" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/panamcs/");

      const { data: result } = await stagehand.extract(
        "extract a list of ALL the staff members on this page, with their name and their job title",
        z.object({
          staff_members: z.array(
            z.object({
              name: z.string(),
              job_title: z.string(),
            }),
          ),
        }),
      );

      const expectedStaff = await page.evaluate(() =>
        Array.from(document.querySelectorAll(".team-tiles .team-tile")).map((tile) => ({
          name: tile.querySelector("h3")?.textContent?.trim() ?? "",
          job_title: tile.querySelector("h4")?.textContent?.trim() ?? "",
        })),
      );
      const key = (member: { name: string; job_title: string }) =>
        `${normalizeString(member.name)}|${normalizeString(member.job_title)}`;
      const actualKeys = result.staff_members.map(key).sort();
      const expectedKeys = expectedStaff.map(key).sort();
      const allStaffMatch =
        actualKeys.length === expectedKeys.length &&
        JSON.stringify(actualKeys) === JSON.stringify(expectedKeys);

      if (!allStaffMatch) {
        logger.error({
          message: "Extracted staff members do not match the page",
          level: 0,
          auxiliary: {
            expected: { value: JSON.stringify(expectedStaff), type: "object" },
            actual: { value: JSON.stringify(result.staff_members), type: "object" },
          },
        });
        return {
          _success: false,
          error: "Extracted staff members do not match the page",
          logs: logger.getLogs(),
          debugUrl,
          sessionUrl,
        };
      }

      return {
        _success: true,
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } catch (error) {
      return {
        _success: false,
        error: error instanceof Error ? error.message : String(error),
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } finally {
      await stagehand.close();
    }
  },
);

File: packages/sdk-ruby/evals/tasks/extract/extract_staff_members.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/extract/extract_staff_members.ts

Evals.define_task("extract_staff_members") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/panamcs/")

  result = t.stagehand.extract(
    "extract a list of ALL the staff members on this page, with their name and their job title",
    schema: {
      "type" => "object",
      "properties" => {
        "staff_members" => {
          "type" => "array",
          "items" => {
            "type" => "object",
            "properties" => {
              "name" => { "type" => "string" },
              "job_title" => { "type" => "string" },
            },
            "required" => %w[name job_title],
            "additionalProperties" => false,
          },
        },
      },
      "required" => ["staff_members"],
      "additionalProperties" => false,
    },
  )

  staff_members = result.data["staff_members"]

  expected_length = 50

  expected_first_item = {
    "name" => "Louis Alvarez",
    "job_title" => "School Resource Officer",
  }

  expected_last_item = {
    "name" => "Jessica Zipin",
    "job_title" => "School Based Therapist",
  }

  if staff_members.length != expected_length
    t.logger.error("Incorrect number of items extracted",
                   { expected: expected_length, actual: staff_members.length })
    next { _success: false, error: "Incorrect number of staff members extracted" }
  end

  # Check for the presence of the expected items
  first_item_exists = staff_members.any? do |member|
    member["name"] == expected_first_item["name"] &&
      member["job_title"] == expected_first_item["job_title"]
  end

  unless first_item_exists
    t.logger.error("Expected first staff member not found in extracted data",
                   { expected: expected_first_item, actual: staff_members })
    next { _success: false, error: "Expected first staff member not found in extracted data" }
  end

  last_item_exists = staff_members.any? do |member|
    member["name"] == expected_last_item["name"] &&
      member["job_title"] == expected_last_item["job_title"]
  end

  unless last_item_exists
    t.logger.error("Expected last staff member not found in extracted data",
                   { expected: expected_last_item, actual: staff_members })
    next { _success: false, error: "Expected last staff member not found in extracted data" }
  end

  { _success: true }
end
